packages/manager/src/features/Lish/Lish.tsx

```diff
@@ -23,10 +23,48 @@ import type { Tab } from 'src/components/Tabs/TabLinkList';
 
 const AUTH_POLLING_INTERVAL = 2000;
 
+export class RetryLimiter {
+  maxTries: number;
+  perTimeWindowMs: number;
+  retryTimes: number[];
+
+  constructor(maxTries: number, perTimeWindowMs: number) {
+    this.maxTries = maxTries;
+    this.perTimeWindowMs = perTimeWindowMs;
+    this.retryTimes = [];
+  }
+  retryAllowed(): boolean {
+    const now = Date.now();
+    this.retryTimes.push(now);
+    const cutOffTime = now-this.perTimeWindowMs;
+    while (this.retryTimes.length && this.retryTimes[0] < cutOffTime)
+      this.retryTimes.shift();
+    return this.retryTimes.length < this.maxTries;
+  }
+  reset(): void {
+    this.retryTimes = [];
+  }
+}
+
+export function formatError(errObj: any, defaultError: string): string {
+  let error = defaultError;
+  if (typeof errObj === 'string') {
+    error = errObj;
+  } else if (errObj?.reason) {
+    error = `${errObj?.reason}`;
+  } else if (errObj?.errors?.[0]?.reason) {
+    error = `Error code: ${errObj.errors[0].reason}`;
+  }
+  if (errObj?.grn) {
+    error = `${error} (${errObj?.grn})`;
+  }
+  return error;
```

Comment:
We can just return without the intermediary error variable
```suggestion
  if (typeof errObj === 'string') {
    return errObj;
  } else if (errObj?.reason) {
    return `${errObj?.reason}`;
  } else if (errObj?.errors?.[0]?.reason) {
    return `Error code: ${errObj.errors[0].reason}`;
  }
  if (errObj?.grn) {
    return `${defaultError} (${errObj?.grn})`;
  }
  return defaultError;
```

Reply:
The last case appends to the error already built. I reorganized the code to use two helper functions that construct the two parts separately. The resulting code is much easier to undestand.